File: packages/typescript/src/utils/validators.ts
/**
 * Client-side validation utilities.
 */

import { ValidationException, InvalidCardException } from '../errors/index.js';
import type { PaymentIntent } from '../types/index.js';

const SUPPORTED_CURRENCIES = ['COP', 'MXN', 'DOP', 'USD', 'EUR'];

const COUNTRY_RULES: Record<
  string,
  {
    documentTypes: string[];
    paymentSystems: number[];
    requiresUserTypeForPSE?: boolean;
    speiRequiresNoCard?: boolean;
    preAuthOnlyFiCode?: string;
  }
> = {
  CO: {
    documentTypes: ['CC', 'NIT', 'PP', 'CE', 'DE'],
    paymentSystems: [0, 1],
    requiresUserTypeForPSE: true,
  },
  DO: {
    documentTypes: ['CI', 'RNC', 'PP'],
    paymentSystems: [3, 6],
    preAuthOnlyFiCode: 'AZUL',
  },
  MX: {
    documentTypes: ['CURP', 'IFE', 'RFC', 'PP'],
    paymentSystems: [1, 7],
    speiRequiresNoCard: true,
  },
};

/**
 * Standard Luhn algorithm to validate card numbers.
 */
export function luhnCheck(cardNumber: string): boolean {
  const digits = cardNumber.replace(/\D/g, '');
  if (digits.length < 13 || digits.length > 19) return false;

  let sum = 0;
  let shouldDouble = false;

  for (let i = digits.length - 1; i >= 0; i--) {
    let digit = parseInt(digits[i], 10);
    if (shouldDouble) {
      digit *= 2;
      if (digit > 9) digit -= 9;
    }
    sum += digit;
    shouldDouble = !shouldDouble;
  }

  return sum % 10 === 0;
}

/**
 * Validate email format.
 */
export function validateEmail(email: string): boolean {
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
}

/**
 * Validate a PaymentIntent before sending to the API.
 */
export function validatePaymentIntent(intent: PaymentIntent, configEtyCode?: number): void {
  if (!intent.amount || intent.amount <= 0) {
    throw new ValidationException('amount must be greater than 0');
  }

  if (!/^\d+(\.\d{1,2})?$/.test(String(intent.amount)) && intent.amount <= 0) {
    throw new ValidationException('amount must be a positive number with at most 2 decimal places');
  }

  if (!intent.currency || !SUPPORTED_CURRENCIES.includes(intent.currency.toUpperCase())) {
    throw new ValidationException(
      `currency "${intent.currency}" is not supported. Supported: ${SUPPORTED_CURRENCIES.join(', ')}`,
    );
  }

  if (!intent.customer) {
    throw new ValidationException('customer is required');
  }

  if (!intent.customer.email || !validateEmail(intent.customer.email)) {
    throw new ValidationException('customer.email is invalid');
  }

  if (!intent.customer.fullName || intent.customer.fullName.trim().length === 0) {
    throw new ValidationException('customer.fullName is required');
  }

  // etyCode must come from either intent or config
  if (configEtyCode === undefined || configEtyCode === null) {
    throw new ValidationException('etyCode is required (set in EcollectClient config)');
  }
}

/**
 * Validate country-specific rules.
 * @param intent - The payment intent
 * @param country - ISO 3166-1 alpha-2, e.g. "CO", "MX", "DO"
 */
export function validateByCountry(intent: PaymentIntent, country: string): void {
  const rules = COUNTRY_RULES[country.toUpperCase()];
  if (!rules) {
    // Unknown country: no specific validations
    return;
  }

  if (
    intent.customer.documentType &&
    !rules.documentTypes.includes(intent.customer.documentType.toUpperCase())
  ) {
    throw new ValidationException(
      `documentType "${intent.customer.documentType}" is not valid for country ${country}. ` +
        `Allowed: ${rules.documentTypes.join(', ')}`,
    );
  }

  const psSystem = intent.paymentSystem ? parseInt(intent.paymentSystem, 10) : undefined;

  if (psSystem !== undefined && !isNaN(psSystem) && !rules.paymentSystems.includes(psSystem)) {
    throw new ValidationException(
      `paymentSystem ${psSystem} is not available in country ${country}. ` +
        `Allowed: ${rules.paymentSystems.join(', ')}`,
    );
  }

  // PSE (paymentSystem=0) requires UserType in Colombia
  if (psSystem === 0 && rules.requiresUserTypeForPSE) {
    if (intent.userType === undefined || intent.userType === null || intent.userType === '') {
      throw new ValidationException(
        'userType is required for PSE payments in Colombia (0=Natural, 1=Juridica)',
      );
    }
  }

  // SPEI (paymentSystem=7) in Mexico cannot have card data
  if (psSystem === 7 && rules.speiRequiresNoCard && intent.tokenId) {
    throw new ValidationException('SPEI payments in Mexico do not support card/token data');
  }
}

/**
 * Validate card data before tokenisation.
 */
export function validateCardNumber(cardNumber: string): void {
  const digits = cardNumber.replace(/\D/g, '');
  if (!luhnCheck(digits)) {
    throw new InvalidCardException(`Card number failed Luhn check`);
  }
}

/**
 * Validate expiration date format MM/YYYY and that it's not in the past.
 */
export function validateExpirationDate(expiry: string): void {
  const match = expiry.match(/^(\d{2})\/(\d{4})$/);
  if (!match) {
    throw new InvalidCardException('expirationDate must be in MM/YYYY format');
  }
  const month = parseInt(match[1], 10);
  const year = parseInt(match[2], 10);
  if (month < 1 || month > 12) {
    throw new InvalidCardException('expirationDate month is out of range');
  }
  const now = new Date();
  const expDate = new Date(year, month - 1, 1);
  const firstOfThisMonth = new Date(now.getFullYear(), now.getMonth(), 1);
  if (expDate < firstOfThisMonth) {
    throw new InvalidCardException('Card has expired');
  }
}

File: packages/python/ecollect/utils/validators.py
"""Validation helpers for payment data."""
import re
from typing import Optional

from ecollect.exceptions import InvalidCardError, ValidationError


def luhn_check(card_number: str) -> bool:
    """Return True if card_number passes the Luhn algorithm.

    Note: all-zeros (e.g. '0000000000000000') is mathematically valid.
    The invalid test number is '4111111111111112' (off-by-one from the
    Visa test number 4111111111111111).
    """
    digits = [int(d) for d in card_number if d.isdigit()]
    if len(digits) < 13:
        return False
    total = 0
    for i, digit in enumerate(reversed(digits)):
        if i % 2 == 1:
            digit *= 2
            if digit > 9:
                digit -= 9
        total += digit
    return total % 10 == 0


def validate_card_number(card_number: str) -> None:
    """Validate a credit card number using the Luhn algorithm."""
    cleaned = re.sub(r"\s+", "", card_number)
    if not cleaned.isdigit():
        raise InvalidCardError("Card number must contain only digits")
    if not luhn_check(cleaned):
        raise InvalidCardError(f"Card number {cleaned!r} failed Luhn check")


def validate_expiration_date(expiry: str) -> None:
    """Validate expiration date format MM/YYYY."""
    if not re.match(r"^\d{2}/\d{4}$", expiry):
        raise ValidationError("Expiration date must be in MM/YYYY format")
    month, year = expiry.split("/")
    if not (1 <= int(month) <= 12):
        raise ValidationError("Expiration month must be between 01 and 12")


def validate_email(email: str) -> None:
    """Basic email format validation."""
    if not re.match(r"^[^\s@]+@[^\s@]+\.[^\s@]+$", email):
        raise ValidationError(f"Invalid email format: {email!r}")


def validate_payment_intent(intent: "PaymentIntent") -> None:  # type: ignore[name-defined]
    """Validate a PaymentIntent before sending to the API."""
    from ecollect.types import PaymentIntent  # avoid circular import

    if intent.amount <= 0:
        raise ValidationError("amount must be greater than 0")
    if not intent.currency:
        raise ValidationError("currency is required")
    if not intent.customer:
        raise ValidationError("customer is required")
    if not intent.customer.email:
        raise ValidationError("customer.email is required")
    validate_email(intent.customer.email)


# Country-specific validation rules
_COUNTRY_RULES = {
    "CO": {
        "document_types": {"CC", "NIT", "PP", "CE", "DE"},
        "payment_systems": {0, 1},
        "requires_user_type_for_pse": True,
    },
    "DO": {
        "document_types": {"CI", "RNC", "PP"},
        "payment_systems": {3, 6},
        "requires_user_type_for_pse": False,
    },
    "MX": {
        "document_types": {"CURP", "IFE", "RFC", "PP"},
        "payment_systems": {1, 7},
        "requires_user_type_for_pse": False,
        "spei_no_card": True,
    },
}


def validate_by_country(intent: "PaymentIntent", country: str) -> None:  # type: ignore[name-defined]
    """Apply country-specific validation rules to a PaymentIntent."""
    rules = _COUNTRY_RULES.get(country.upper())
    if not rules:
        return  # unknown country — skip validation

    doc_type = intent.customer.document_type
    if doc_type and doc_type not in rules["document_types"]:
        raise ValidationError(
            f"DocumentType {doc_type!r} is not valid for country {country}"
        )

    ps = intent.payment_system
    if ps is not None and ps not in rules["payment_systems"]:
        raise ValidationError(
            f"PaymentSystem {ps} is not available in country {country}"
        )

    # Colombia PSE (PaymentSystem=0) requires UserType
    if country.upper() == "CO" and ps == 0 and not intent.user_type:
        raise ValidationError("UserType is required for PSE (Colombia)")

    # Mexico SPEI (PaymentSystem=7) cannot use card data
    if country.upper() == "MX" and ps == 7 and intent.token_id:
        raise ValidationError("SPEI (Mexico) does not support card/token data")
